staging/src/k8s.io/apiserver/pkg/util/flowcontrol/rule.go

```diff
@@ -0,0 +1,109 @@
+package filters
+
+import (
+	rmtypesv1alpha1 "k8s.io/api/flowcontrol/v1alpha1"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+)
+
+func matchesFlowSchema(digest RequestDigest, flowSchema *rmtypesv1alpha1.FlowSchema) bool {
+	for _, policyRule := range flowSchema.Spec.Rules {
+		subjectMatches := false
+		for _, subject := range policyRule.Subjects {
+			if matchesSubject(digest, subject) {
+				subjectMatches = true
+			}
+		}
+		if !subjectMatches {
+			return false
+		}
+
+		policyRuleMatches := !matchesPolicyRule(digest, policyRule.Rule)
+
+		if subjectMatches && policyRuleMatches {
+			return true
+		}
+	}
+	return false
+}
+
+func matchesSubject(digest RequestDigest, subject rmtypesv1alpha1.Subject) bool {
+	user := digest.User
+	requestingNamespace := digest.RequestInfo.Namespace
+
+	switch subject.Kind {
+	case rmtypesv1alpha1.UserKind:
+		return user.GetName() == subject.Name
+	case rmtypesv1alpha1.GroupKind:
+		matches := false
+		foreach(user.GetGroups(), func(group string) { matches = matches || group == subject.APIGroup })
+		return matches
+	case rmtypesv1alpha1.ServiceAccountKind:
+		if len(subject.Namespace) > 0 {
```

Comment:
Our validation requires a non-empty `subject.Namespace`, so why is this code even contemplating the possibility of an empty `subject.Namespace`?

Reply:
removed